### PR TITLE
[FLINK-10295] Add support of passing jar arguments as list of separate strings in REST API

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

This PR add one more option to specify jar program arguments in REST API calls (run and get plan). Users can specify arguments as one string or additionally mow as a list of a separate strings.

## Brief change log

  - add JarRequestBody as a base for JarRunRequestBody and JarPlanRequestBody
  - add programArgsList to JarRequestBody
  - add JarHandlerUtils.getProgramArgs to resolve program args from programArgs and programArgsList from query path or json body in JarRunHandler or JarPlanHandler
  - user can use programArgs or programArgsList but not both, add test for it

## Verifying this change

Unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by azagrebin** [comment]
cc @zentol

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandlerParameterTest.java [hunk 1]
```diff
@@ -219,15 +225,40 @@ public void testConfigurationViaJsonRequest() throws Exception {
...
+		try {
+			testParameterPrioritization(PROG_ARGS, PROG_ARGS);
+			fail("IllegalArgumentException is excepted");
+		} catch (IllegalArgumentException e) {
+			// expected
+		}
+	}
+
+	private void testParameterPrioritization(List<String> programArgs, List<String> programArgsList) throws Exception {
+		List<String> args = programArgs == null ? programArgsList : programArgs;
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandlerParameterTest.java]: This code is way to confusing. It strikes me as very odd that `args` is never null, and the naming scheme is weird (args should be "argsList", "argsStr" should be "args" to be somewhat consistent with the method arguments).

I would pass in a single collection as arguments and a simple enum for how arguments should be passed.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/utils/JarHandlerUtils.java [hunk 2]
```diff
@@ -26,13 +36,42 @@
...
+				null,
+				log));
+		List<String> programArgsList = fromRequestBodyOrQueryParameter(
+			requestBody.getProgramArgumentsList(),
+			() -> request.getQueryParameter(ProgramArgsListQueryParameter.class),
+			null,
+			log);
+		if (requestBody.getProgramArgumentsList() != null) {
+			if (!programArgs.isEmpty()) {
+				throw new IllegalArgumentException(
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/utils/JarHandlerUtils.java]: neither handler is handling this exception, so the user will get an INTERNAL_SERVER_ERROR instead of BAD_REQUEST.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRequestBody.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+
+/**
+ * Base class for {@link RequestBody} for running a jar or querying the plan.
+ */
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public abstract class JarRequestBody implements RequestBody {
```
[the remark this answers is not in the document]
Comment (azagrebin) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRequestBody.java]: constructor and `getEntryClassName`, the class is used in `JarHandlerUtils.getProgramArgs` to resolve program args

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandlerParameterTest.java [hunk 4]
```diff
@@ -57,17 +61,27 @@
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
 
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.fail;
+
 /**
  * Tests for the parameter handling of the {@link JarRunHandler}.
  */
 public class JarRunHandlerParameterTest extends TestLogger {
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandlerParameterTest.java]: A similar test is required for the `JarPlanHandler`.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsListQueryParameter.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+import java.io.File;
+
+/**
+ * Query parameter specifying the arguments list for the program.
+ * @see org.apache.flink.client.program.PackagedProgram#PackagedProgram(File, String, String...)
+ */
+public class ProgramArgsListQueryParameter extends StringQueryParameter {
+	public ProgramArgsListQueryParameter() {
+		super("program-args-list", MessageParameter.MessageParameterRequisiteness.OPTIONAL);
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsListQueryParameter.java]: Rename to `programArgsList` to adhere to the new naming scheme, and use `FIELD_NAME_PROGRAM_ARGUMENTS_LIST`.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsQueryParameter.java [hunk 6]
```diff
@@ -32,6 +32,7 @@ public ProgramArgsQueryParameter() {
 
 	@Override
 	public String getDescription() {
-		return "String value that specifies the arguments for the program or plan.";
+		return "Deprecated, please, use 'programArgsList' instead. " +
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsQueryParameter.java]: Let's also add a `@Deprecated` annotation to the class.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsQueryParameter.java [hunk 7]
```diff
@@ -32,6 +32,7 @@ public ProgramArgsQueryParameter() {
...
+			"String value that specifies the arguments for the program or plan.";
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsQueryParameter.java]: Remove comma after please.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsQueryParameter.java [hunk 8]
```diff
@@ -32,6 +32,7 @@ public ProgramArgsQueryParameter() {
 
 	@Override
 	public String getDescription() {
-		return "String value that specifies the arguments for the program or plan.";
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsQueryParameter.java]: Replace `programArgsList` with `JarRequestBody#FIELD_NAME_PROGRAM_ARGUMENTS_LIST`.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsListQueryParameter.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+import org.apache.flink.runtime.rest.messages.MessageParameter;
+
+import java.io.File;
+
+/**
+ * Query parameter specifying one or more arguments for the program.
+ * @see org.apache.flink.client.program.PackagedProgram#PackagedProgram(File, String, String...)
+ */
+public class ProgramArgsListQueryParameter extends StringQueryParameter {
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsListQueryParameter.java]: Rename to `ProgramArgQueryParameter`

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+
+	static List<String> getProgramArgsList(ProgramArgsParType programArgsParType) {
+		return programArgsParType == ProgramArgsParType.List || programArgsParType == ProgramArgsParType.Both
+			? Arrays.asList(PROG_ARGS) : null;
+	}
+
+	private static <
+		REQB extends JarRequestBody,
+		M extends JarMessageParameters>
+	HandlerRequest<REQB, M> createRequest(
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java]: the indentation doesn't look correct, it should not be in the same column as `private static <`

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandlerParameterTest.java [hunk 11]
```diff
@@ -118,197 +68,87 @@ public static void setup() throws Exception {
 			executor);
 	}
 
-	@Before
-	public void reset() {
-		ParameterProgram.actualArguments = null;
+	@Override
+	JarRunMessageParameters getUnresolvedJarMessageParameters() {
+		return JarRunHeaders.getInstance().getUnresolvedMessageParameters();
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandlerParameterTest.java]: use `handler.getMessageHeaders().getUnresolvedMessageParameters()` instead

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarPlanHandlerParameterTest.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+			executor,
+			jobGraph -> {
+				LAST_SUBMITTED_JOB_GRAPH_REFERENCE.set(jobGraph);
+				return new JobPlanInfo(JsonPlanGenerator.generatePlan(jobGraph));
+			});
+	}
+
+	@Override
+	JarPlanMessageParameters getUnresolvedJarMessageParameters() {
+		return JarPlanHeaders.getInstance().getUnresolvedMessageParameters();
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarPlanHandlerParameterTest.java]: use `handler.getMessageHeaders().getUnresolvedMessageParameters()` instead

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarPlanHandlerParameterTest.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+			localAddressFuture,
+			gatewayRetriever,
+			timeout,
+			responseHeaders,
+			JarPlanHeaders.getInstance(),
+			jarDir,
+			new Configuration(),
+			executor,
+			jobGraph -> {
+				LAST_SUBMITTED_JOB_GRAPH_REFERENCE.set(jobGraph);
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarPlanHandlerParameterTest.java]: why is this necessary? Which case does this cover that the `submitFunction` of the `TestingDispatcherGateway` doesn't?

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarPlanHandlerParameterTest.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+	@Override
+	JarPlanMessageParameters getWrongJarMessageParameters(ProgramArgsParType programArgsParType) {
+		List<String> wrongArgs = Arrays.stream(PROG_ARGS).map(a -> a + "wrong").collect(Collectors.toList());
+		String argsWrongStr = String.join(" ", wrongArgs);
+		final JarPlanMessageParameters parameters = getUnresolvedJarMessageParameters();
+		parameters.entryClassQueryParameter.resolve(Collections.singletonList("please.dont.run.me"));
+		parameters.parallelismQueryParameter.resolve(Collections.singletonList(64));
+		if (programArgsParType == ProgramArgsParType.String ||
+			programArgsParType == ProgramArgsParType.Both) {
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarPlanHandlerParameterTest.java]: let's move this to the previous; every time i scan over this line I believe it to not be part of the condition (also applies to other instances in #getJarMessageParameters)

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static junit.framework.TestCase.assertEquals;
+import static junit.framework.TestCase.fail;
+
+/** Base test class for jar request handlers. */
+public abstract class JarHandlerParameterTest
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java]: package-private?

File: docs/_includes/generated/rest_v1_dispatcher.html [hunk 16]
```diff
@@ -290,19 +290,40 @@
     <tr>
       <td colspan="2">
         <ul>
+<li><code>program-args</code> (optional): Deprecated, please, use 'programArgsList' in a JSON request body as a JSON list. String value that specifies the arguments for the program or plan.</li>
```
Comment (zentol) [inline docs/_includes/generated/rest_v1_dispatcher.html]: needs regeneration

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarPlanHandlerParameterTest.java (hunk 13 above)
Comment (azagrebin) [reply flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarPlanHandlerParameterTest.java]: the jar plan handler does not call `submitFunction`, it just generates the plan.
